### PR TITLE
Add LSTM with timegate

### PR DESCRIPTION
**Pull request by fferroni**
Adding a timegate to the LSTM architecture showed great promise at this year's NIPS conference. 
A simple example on MNIST dataset is shown here, compared to standard LSTM in Keras: https://github.com/fferroni/PhasedLSTM-Keras/blob/master/README.md 

I am wondering if it could become part of the Keras library, and if so, what are the requirements for the code to be accepted.

Thank you for the great library.

### CODE REVIEW

**Comment by tboquet** [comment]
Cool work, is your implementation working with both Theano and TensorFlow?

**Comment by fferroni** [comment]
Hi, 

At the moment it does not work on Tensorflow because the K.switch() implementation for the Theano backend accepts tensors, but Tensorflow backend does not (relevant discussion found here https://github.com/fchollet/keras/issues/4725 and here https://github.com/fchollet/keras/pull/4519).

Either I use a different implementation for TF, or just wait until K.switch() functionality is extended for the Tensorflow backend to match that of the Theano backend. For the prior, I would need some help because I am more familiar with Theano.

**Comment by tboquet** [comment]
I have something working with TensorFlow based on your code. I also implemented the layer on my side after NIPS so I only copy pasted a small part :smile: . I made small changes though and I use masks to have the switch behavior. I'm not sure it's really efficient but it works. One quick question:  does the backward pass compile with Theano with the modulo operator and does your code run as is with Theano? (the gradient of the mod is not implemented in TensorFlow)

**Comment by fferroni** [comment]
Yup, on Theano works just fine as it, you can see it from the README.md

For Tensorflow, the gradient of the mod would look something like:
```
@ops.RegisterGradient("Mod")
def _mod_grad(op, grad):
    x, y = op.inputs
    gz = grad
    x_grad = gz
    y_grad = tf.reduce_mean(-(x // y) * gz, reduction_indices=[0], keep_dims=True)
    return x_grad, y_grad
```

**Comment by tboquet** [comment]
Do you plan to open a PR in TF?

I used this hack to have switch behavior:

```python
up = K.cast(K.lesser(phi, r_on * 0.5), 'float32') * 2 * phi / r_on
mid = K.cast(K.lesser(phi, r_on), 'float32') * K.cast(K.greater(phi, r_on * 0.5), 'float32') * (2 - (2 * phi / r_on))
end = K.cast(K.greater(phi, r_on * 0.5), 'float32') * self.alpha * phi
k = up + mid + end
```

where:

```python
r_on = self.timegate[2]
```

From you code.

**Comment by fchollet** [comment]
We won't be adding new layers into core Keras for the time being.

This layer would be a better fit for a 3rd-party repository of Keras extensions.

**Comment by fferroni** [comment]
Ok, I understand.

@tboquet, thanks for the tips. I made some modifications (can be found in original repo), and now working for both for Theano and Tensorflow.
